addons/stock/models/stock_picking.py

```diff
@@ -948,44 +948,43 @@ def _check_entire_pack(self):
+            lines_to_check_ids = [line_id for picking in self for line_id in get_relevant_move_line_ids(no_qty_done_dict, picking)]
+        else:
+            lines_to_check_ids = self.move_line_ids.ids
+            if any(not no_qty_done_dict.get(picking, True) for picking in self):
+                lines_to_check_ids = get_line_with_done_qty_ids(self.move_line_ids)
+
+        return self.env['stock.move.line'].browse(lines_to_check_ids)
+
+    def _sanity_check(self, separate_pickings=True):
+        """ Sanity check for `button_validate()`
+            :param separate_pickings: Indicates if pickings should be checked independantly for lot/serial numbers or not.
+        """
         pickings_without_lots = self.browse()
         products_without_lots = self.env['product.product']
-        for picking in self:
-            if not picking.move_ids and not picking.move_line_ids:
-                pickings_without_moves |= picking
+        pickings_without_moves = self.filtered(lambda p: not p.move_ids and not p.move_line_ids)
+        precision_digits = self.env['decimal.precision'].precision_get('Product Unit of Measure')
 
-            picking.message_subscribe([self.env.user.partner_id.id])
-            picking_type = picking.picking_type_id
-            precision_digits = self.env['decimal.precision'].precision_get('Product Unit of Measure')
-            no_quantities_done = all(float_is_zero(move_line.qty_done, precision_digits=precision_digits) for move_line in picking.move_line_ids.filtered(lambda m: m.state not in ('done', 'cancel')))
-            no_reserved_quantities = all(float_is_zero(move_line.reserved_qty, precision_rounding=move_line.product_uom_id.rounding) for move_line in picking.move_line_ids)
-            if no_reserved_quantities and no_quantities_done:
-                pickings_without_quantities |= picking
-
-            if picking_type.use_create_lots or picking_type.use_existing_lots:
-                lines_to_check = picking.move_line_ids
-                if not no_quantities_done:
-                    lines_to_check = lines_to_check.filtered(lambda line: float_compare(line.qty_done, 0, precision_rounding=line.product_uom_id.rounding))
-                for line in lines_to_check:
-                    product = line.product_id
-                    if product and product.tracking != 'none':
-                        if not line.lot_name and not line.lot_id:
-                            pickings_without_lots |= picking
-                            products_without_lots |= product
+        no_quantities_done = {picking: all(float_is_zero(move_line.qty_done, precision_digits=precision_digits) for move_line in picking.move_line_ids.filtered(lambda m: m.state not in ('done', 'cancel'))) for picking in self}
+        no_reserved_quantities = {picking: all(float_is_zero(move_line.reserved_qty, precision_rounding=move_line.product_uom_id.rounding) for move_line in picking.move_line_ids) for picking in self}
```

Comment:
Since the value is just `True/False`, maybe `set` is better than `dict`

Reply:
You might be right.
At least it makes the comparisons further down much clearer when checking on a set rather than a dict, so I'll change it.